File: aoc/2024/aoc202412.hs
module AOC202412 where

import Data.Array(Array,assocs,bounds,inRange,(!))
import Data.Set(Set,empty,insert,member)

import AOC

aoc = AOC {
    day="12",
    aocTests=[
        AOCTest {
            testData=unlines [
                "AAAA",
                "BBCD",
                "BBCC",
                "EEEC"
                ],
            testResult=Just "140",
            testResult2=Just "80"
            },
        AOCTest {
            testData=unlines [
                "OOOOO",
                "OXOXO",
                "OOOOO",
                "OXOXO",
                "OOOOO"
                ],
            testResult=Just "772",
            testResult2=Just "436"
            },
        AOCTest {
            testData=unlines [
                "AAAAAA",
                "AAABBA",
                "AAABBA",
                "ABBAAA",
                "ABBAAA",
                "AAAAAA"
                ],
            testResult=Nothing,
            testResult2=Just "368"
            },
        AOCTest {
            testData=unlines [
                "RRRRIICCFF",
                "RRRRIICCCF",
                "VVRRRCCFFF",
                "VVRCCCJFFF",
                "VVVVCJJCFE",
                "VVIVCCJJEE",
                "VVIIICJJEE",
                "MIIIIIJJEE",
                "MIIISIJEEE",
                "MMMISSJEEE"
                ],
            testResult=Just "1930",
            testResult2=Just "1206"
            }
        ],
    aocCode=Code {
        codeParse=parse2da,
        codeParse2=parse2da,
        codeTest=result perimeter,
        codeTest2=result sides,
        codeResult=result perimeter,
        codeResult2=result sides
        }
    }

measureRegions :: (Array (Int,Int) Char -> Char -> (Int,Int) -> Int)
               -> Array (Int,Int) Char -> [(Int,Int)]
measureRegions metric grid = snd $ foldr measure (empty,[]) $ assocs grid
  where
    inBounds = inRange (bounds grid)
    measure (xy,ch) (seen,results)
      | member xy seen = (seen,results)
      | otherwise = fmap (:results)
                      $ foldr (fill ch) (insert xy seen,(1,metric grid ch xy))
                      $ filter inBounds $ neighbors xy
    fill ch xy (seen,(count,m))
      | grid!xy /= ch || member xy seen = (seen,(count,m))
      | otherwise =
          foldr (fill ch) (insert xy seen,(1+count,m+metric grid ch xy))
              $ filter inBounds $ neighbors xy

perimeter :: Array (Int,Int) Char -> Char -> (Int,Int) -> Int
perimeter regions region xy =
    length [() | n <- neighbors xy, not (inBounds n && regions!n == region)]
  where
    inBounds = inRange (bounds regions)

-- Counts south ends of fences to the east, north ends of fences to the west,
-- east ends of fences to the south, and west ends of fences to the north.
sides :: Array (Int,Int) Char -> Char -> (Int,Int) -> Int
sides regions region (x,y) =
    length [() | dxy <- [(1,0),(-1,0),(0,1),(0,-1)], isEnd dxy]
  where
    inRegion (dx,dy) = inRange (bounds regions) (x+dx,y+dy)
                            && region == regions!(x+dx,y+dy)
    isEnd (dx,dy) = not (inRegion (dx,dy))
        && (inRegion (dy,dx),inRegion (dx+dy,dy+dx)) /= (True,False)

neighbors :: (Int,Int) -> [(Int,Int)]
neighbors (x,y) = [(x+1,y),(x-1,y),(x,y+1),(x,y-1)]

result metric = sum . map (uncurry (*)) . measureRegions metric

File: aoc/2025/aoc202412.go
package main

import (
	"container/list"
)

func init() {
	Register(&aoc202412{
		AOC: AOC{
			Day:           12,
			InputFilename: "../2024/input/12.txt",
			Tests: []Test{
				Test{
					Input: `AAAA
BBCD
BBCC
EEEC
`,
					Part1: "140",
					Part2: "80",
				},
				Test{
					Input: `OOOOO
OXOXO
OOOOO
OXOXO
OOOOO
`,
					Part1: "772",
					Part2: "436",
				},
				Test{
					Input: `RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE
`,
					Part1: "1930",
					Part2: "1206",
				},
				Test{
					Input: `AAAAAA
AAABBA
AAABBA
ABBAAA
ABBAAA
AAAAAA
`,
					Part1: "",
					Part2: "368",
				},
			},
		},
	})
}

type aoc202412 struct {
	AOC
}

func (aoc *aoc202412) markRegion(grid map[[2]int]byte, xy [2]int, marks map[[2]int]bool) map[[2]int][2]int {
	region := map[[2]int][2]int{}
	regionid := grid[xy]
	queue := list.New()
	queue.PushFront(xy)
	for {
		e := queue.Front()
		if e == nil {
			return region
		}
		queue.Remove(e)
		xy := e.Value.([2]int)
		if marks[xy] {
			continue
		}
		marks[xy] = true
		fences := 4
		leftFence := true
		if regionid == grid[[2]int{xy[0] - 1, xy[1]}] {
			fences--
			leftFence = false
			if !marks[[2]int{xy[0] - 1, xy[1]}] {
				queue.PushFront([2]int{xy[0] - 1, xy[1]})
			}
		}
		rightFence := true
		if regionid == grid[[2]int{xy[0] + 1, xy[1]}] {
			fences--
			rightFence = false
			if !marks[[2]int{xy[0] + 1, xy[1]}] {
				queue.PushFront([2]int{xy[0] + 1, xy[1]})
			}
		}
		upFence := true
		if regionid == grid[[2]int{xy[0], xy[1] - 1}] {
			fences--
			upFence = false
			if !marks[[2]int{xy[0], xy[1] - 1}] {
				queue.PushFront([2]int{xy[0], xy[1] - 1})
			}
		}
		downFence := true
		if regionid == grid[[2]int{xy[0], xy[1] + 1}] {
			fences--
			downFence = false
			if !marks[[2]int{xy[0], xy[1] + 1}] {
				queue.PushFront([2]int{xy[0], xy[1] + 1})
			}
		}
		fenceEnds := 0
		if leftFence {
			if upFence {
				fenceEnds++
			} else if regionid == grid[[2]int{xy[0] - 1, xy[1] - 1}] {
				fenceEnds++
			}
		}
		if rightFence {
			if upFence {
				fenceEnds++
			} else if regionid == grid[[2]int{xy[0] + 1, xy[1] - 1}] {
				fenceEnds++
			}
		}
		if upFence {
			if leftFence {
				fenceEnds++
			} else if regionid == grid[[2]int{xy[0] - 1, xy[1] - 1}] {
				fenceEnds++
			}
		}
		if downFence {
			if leftFence {
				fenceEnds++
			} else if regionid == grid[[2]int{xy[0] - 1, xy[1] + 1}] {
				fenceEnds++
			}
		}
		region[xy] = [2]int{fences, fenceEnds}
	}
}

func (aoc *aoc202412) parse(input *Input) []map[[2]int][2]int {
	_, _, grid := input.Grid()
	var regions []map[[2]int][2]int
	marks := map[[2]int]bool{}
	for xy := range grid {
		if marks[xy] {
			continue
		}
		regions = append(regions, aoc.markRegion(grid, xy, marks))
	}

	return regions
}

func (aoc *aoc202412) cost(region map[[2]int][2]int) int {
	area := len(region)
	perimeter := 0
	for _, plot := range region {
		perimeter += plot[0]
	}
	return area * perimeter
}

func (aoc *aoc202412) Part1(input *Input) string {
	result := 0
	for _, region := range aoc.parse(input) {
		result += aoc.cost(region)
	}
	return IntResult(result)
}

func (aoc *aoc202412) bulkCost(region map[[2]int][2]int) int {
	area := len(region)
	sides := 0
	for _, plot := range region {
		sides += plot[1]
	}
	return area * sides
}

func (aoc *aoc202412) Part2(input *Input) string {
	result := 0
	for _, region := range aoc.parse(input) {
		result += aoc.bulkCost(region)
	}
	return IntResult(result)
}
